File: Java/opentraceability/src/main/java/opentraceability/mappers/epcis/xml/EPCISXmlMasterDataReader.java
package opentraceability.mappers.epcis.xml;

import opentraceability.interfaces.*;
import opentraceability.models.common.*;
import opentraceability.models.events.*;
import opentraceability.models.events.kdes.*;
import opentraceability.models.identifiers.*;
import opentraceability.models.masterdata.*;
import opentraceability.models.masterdata.kdes.MasterDataKDEObject;
import opentraceability.models.masterdata.kdes.MasterDataKDEString;
import opentraceability.utility.*;
import opentraceability.utility.attributes.*;
import opentraceability.*;
import opentraceability.mappers.*;
import org.apache.commons.codec.language.bm.Lang;
import org.json.JSONObject;

import javax.xml.xpath.XPathExpressionException;
import java.lang.reflect.Field;

import java.net.URI;
import java.util.*;
import java.util.stream.Collectors;

public final class EPCISXmlMasterDataReader
{
	public static void ReadMasterData(EPCISBaseDocument doc, XElement xMasterData) throws Exception {
		//<VocabularyList>
		//        <Vocabulary type="urn:epcglobal:epcis:vtype:EPCClass">
		//        <Vocabulary type="urn:epcglobal:epcis:vtype:Location">
		//        <Vocabulary type="urn:epcglobal:epcis:vtype:Party">

		XElement xVocabList = xMasterData.Element("VocabularyList");
		if (!xVocabList.IsNull)
		{
			for (XElement xVocab : xVocabList.Elements())
			{
				String type = xVocab.Attribute("type").toLowerCase();
				if (type != null)
				{
					XElement xVocabElementaryList = xVocab.Element("VocabularyElementList");
					if (!xVocabElementaryList.IsNull)
					{
						for (XElement xVocabElement : xVocabElementaryList.Elements())
						{
							switch (type)
							{
								case "urn:epcglobal:epcis:vtype:epcclass":
									ReadTradeitem(doc, xVocabElement, type);
									break;
								case "urn:epcglobal:epcis:vtype:location":
									ReadLocation(doc, xVocabElement, type);
									break;
								case "urn:epcglobal:epcis:vtype:party":
									ReadTradingParty(doc, xVocabElement, type);
									break;
								default:
									ReadUnknown(doc, xVocabElement, type);
									break;
							}
						}
					}
				}
			}
		}
	}

	private static void ReadTradeitem(EPCISBaseDocument doc, XElement xTradeitem, String type) throws Exception {
		Class javaType = Setup.MasterDataTypes.get(type);
		if (javaType == null)
		{
			throw new Exception("Failed to read java type.");
		}

		Object o = ReflectionUtility.constructType(javaType);
		if (o instanceof  TradeItem)
		{
			TradeItem md = (TradeItem)o;
			md.gtin = new opentraceability.models.identifiers.GTIN(xTradeitem.Attribute("id"));
			md.epcisType = type;

			// read the object
			ReadMasterDataObject(md, xTradeitem);
			doc.masterData.add(md);
		}
		else
		{
			throw new Exception(javaType.getTypeName() + " is not instance of TradeItem.");
		}
	}

	private static void ReadLocation(EPCISBaseDocument doc, XElement xLocation, String type) throws Exception {
		Class javaType = Setup.MasterDataTypes.get(type);
		if (javaType == null)
		{
			throw new Exception("Failed to read java type.");
		}

		Object o = ReflectionUtility.constructType(javaType);
		if (o instanceof  Location)
		{
			Location md = (Location)o;
			md.gln = new opentraceability.models.identifiers.GLN(xLocation.Attribute("id"));
			md.epcisType = type;

			// read the object
			ReadMasterDataObject(md, xLocation);
			doc.masterData.add(md);
		}
		else
		{
			throw new Exception(javaType.getTypeName() + " is not instance of Location.");
		}
	}

	private static void ReadTradingParty(EPCISBaseDocument doc, XElement xTradingParty, String type) throws Exception {
		Class javaType = Setup.MasterDataTypes.get(type);
		if (javaType == null)
		{
			throw new Exception("Failed to read java type.");
		}

		Object o = ReflectionUtility.constructType(javaType);
		if (o instanceof  TradingParty)
		{
			TradingParty md = (TradingParty)o;
			md.pgln = new opentraceability.models.identifiers.PGLN(xTradingParty.Attribute("id"));
			md.epcisType = type;

			// read the object
			ReadMasterDataObject(md, xTradingParty);
			doc.masterData.add(md);
		}
		else
		{
			throw new Exception(javaType.getTypeName() + " is not instance of TradingParty.");
		}
	}

	private static void ReadUnknown(EPCISBaseDocument doc, XElement xUnknown, String type) throws Exception {
		// read the pgln from the id
		String id = xUnknown.Element("id").getValue();
		VocabularyElement ele = new VocabularyElement();
		ele.setId(id);
		ele.epcisType = type;

		// read the object
		ReadMasterDataObject(ele, xUnknown);
		doc.masterData.add(ele);
	}


	private static void ReadMasterDataObject(IVocabularyElement md, XElement xMasterData) throws Exception {
		ReadMasterDataObject(md, xMasterData, true);
	}

	private static void ReadMasterDataObject(IVocabularyElement md, XElement xMasterData, boolean readKDEs) throws Exception {
		var mappedProperties = OTMappingTypeInformation.getMasterDataXmlTypeInfo(md.getClass());

		// work on expanded objects...
		// these are objects on the vocab element represented by one or more attributes in the EPCIS master data
		ArrayList<String> ignoreAttributes = new ArrayList<String>();
		for (var property : mappedProperties.properties.stream().filter(p -> p.name.equals("")).collect(Collectors.toList()))
		{
			var subMappedProperties = OTMappingTypeInformation.getMasterDataXmlTypeInfo(property.field.getType());
			boolean setAttribute = false;

			Object subObject = ReflectionUtility.constructType(property.field.getType());
			if (subObject != null)
			{
				for (XElement xeAtt : xMasterData.Elements("attribute"))
				{
					String id = xeAtt.Attribute("id");
					var propMapping = subMappedProperties.get(id, null);
					if (propMapping != null)
					{
						if (!TrySetValueType(xeAtt.getValue(), subObject, propMapping.field, propMapping.itemType))
						{
							Object value = ReadKDEObject(xeAtt, propMapping.field.getType(), propMapping.itemType);
							propMapping.field.set(subObject, value);
						}
						setAttribute = true;
						ignoreAttributes.add(id);
					}
				}
				if (setAttribute)
				{
					property.field.set(md, subObject);
				}
			}
		}

		// go through each standard attribute...
		for (XElement xeAtt : xMasterData.Elements("attribute"))
		{
			String id = xeAtt.Attribute("id");

			if (ignoreAttributes.contains(id))
			{
				continue;
			}

			var propMapping = mappedProperties.get(id, null);
			if (propMapping != null)
			{
				if (!TrySetValueType(xeAtt.getValue(), md, propMapping.field, propMapping.itemType))
				{
					Object value = ReadKDEObject(xeAtt, propMapping.field.getType(), propMapping.itemType);
					propMapping.field.set(md, value);
				}
			}
			else if (readKDEs)
			{
				if (xeAtt.HasElements())
				{
					// serialize into object kde...
					IMasterDataKDE kdeObject = new MasterDataKDEObject("", id);
					kdeObject.setFromEPCISXml(xeAtt);
					md.kdes.add(kdeObject);
				}
				else
				{
					// serialize into string kde
					IMasterDataKDE kdeString = new MasterDataKDEString("", id);
					kdeString.setFromEPCISXml(xeAtt);
					md.kdes.add(kdeString);
				}
			}
		}
	}

	private static Object ReadKDEObject(XElement xeAtt, Class t, Class itemType) throws Exception {
		Object value = ReflectionUtility.constructType(t);

		if (value instanceof List)
		{
			List list = (List)value;
			for (XElement xchild : xeAtt.Elements())
			{
				if (itemType == null)
				{
					throw new Exception("Failed to read generic item type of list.");
				}

				Object child = ReadKDEObject(xchild, itemType, null);
				list.add(child);
			}
		}
		else
		{
			// go through each property...
			for (Field p : t.getClass().getFields())
			{
				OpenTraceabilityAttribute xmlAtt = ReflectionUtility.getFieldAnnotation(p, OpenTraceabilityAttribute.class);
				if (xmlAtt != null)
				{
					XElement x = xeAtt.Element(xmlAtt.name());
					if (x != null)
					{
						OpenTraceabilityObjectAttribute objAtt = ReflectionUtility.getFieldAnnotation(p, OpenTraceabilityObjectAttribute.class);
						if (objAtt != null)
						{
							Object o = ReadKDEObject(x, p.getType(), null);
						}
						else if (!TrySetValueType(x.getValue(), value, p, itemType))
						{
							throw new RuntimeException(String.format("Failed to set value type while reading KDE object. property = %1$s, type = %2$s, xml = %3$s", p.getName(), t.getTypeName(), x));
						}
					}
				}
			}
		}

		return value;
	}

	private static boolean TrySetValueType(String val, Object o, Field p, Class itemType) throws IllegalAccessException {
		if (p.getType() == String.class)
		{
			p.set(o, val);
			return true;
		}
		else if (List.class.isAssignableFrom(p.getType()) && itemType == String.class)
		{
			Object tempVar = p.get(o);
			ArrayList cur = tempVar instanceof ArrayList ? (ArrayList)tempVar : null;
			if (cur == null || cur.isEmpty())
			{
				cur = new ArrayList<String>();
				p.set(o, cur);
			}
			cur.add(val);
			return true;
		}
		else if (p.getType() == Boolean.class || p.getType() == Boolean.class)
		{
			boolean v = Boolean.parseBoolean(val);
			p.set(o, v);
			return true;
		}
		else if (p.getType() == Double.class || p.getType() == Double.class)
		{
			double v = Double.parseDouble(val);
			p.set(o, v);
			return true;
		}
		else if (p.getType() == URI.class)
		{
			URI v = URI.create(val);
			p.set(o, v);
			return true;
		}
		else if (List.class.isAssignableFrom(p.getType()) && itemType == LanguageString.class)
		{
			ArrayList<LanguageString> l = LanguageString.fromJSON(val);
			p.set(o, l);
			return true;
		}
		else if (p.getType() == Country.class)
		{
			Country v = Countries.parse(val);
			p.set(o, v);
			return true;
		}
		else if (p.getType() == PGLN.class)
		{
			PGLN v = new PGLN(val);
			p.set(o, v);
			return true;
		}
		else
		{
			return false;
		}
	}
}

File: Java/OpenTraceability/src/main/opentraceability/mappers/epcis/xml/epcisxmlmasterdatareader.kt
package mappers.epcis.xml

import interfaces.IVocabularyElement
import mappers.OTMappingTypeInformation
import models.common.LanguageString
import javax.xml.bind.annotation.*
import models.events.*
import models.events.EPCISBaseDocument
import models.identifiers.PGLN
import models.masterdata.TradingParty
import models.masterdata.VocabularyElement
import models.masterdata.kdes.MasterDataKDEObject
import models.masterdata.kdes.MasterDataKDEString
import utility.Countries
import utility.Country
import utility.attributes.OpenTraceabilityAttribute
import utility.attributes.OpenTraceabilityObjectAttribute
import java.lang.reflect.Type
import java.net.URI

class EPCISXmlMasterDataReader {
    companion object {

        fun ReadMasterData(doc: EPCISBaseDocument, xMasterData: XElement) {
            val xVocabList = xMasterData.element("VocabularyList")
            if (xVocabList != null) {
                for (xVocab in xVocabList.elements()) {
                    val type = xVocab.attribute("type")?.value?.toLowerCase()
                    if (type != null) {
                        val xVocabElementaryList = xVocab.element("VocabularyElementList")
                        if (xVocabElementaryList != null) {
                            for (xVocabElement in xVocabElementaryList.elements()) {
                                when (type) {
                                    "urn:epcglobal:epcis:vtype:epcclass" -> ReadTradeitem(doc, xVocabElement, type)
                                    "urn:epcglobal:epcis:vtype:location" -> ReadLocation(doc, xVocabElement, type)
                                    "urn:epcglobal:epcis:vtype:party" -> ReadTradingParty(doc, xVocabElement, type)
                                    else -> ReadUnknown(doc, xVocabElement, type)
                                }
                            }
                        }
                    }
                }
            }
        }


        private fun ReadTradeitem(doc: EPCISBaseDocument, xTradeitem: XElement, type: String) {
            // read the GTIN from the id
            val id = xTradeitem.attribute("id")?.value ?: ""
            val tradeitem = Tradeitem()
            tradeitem.GTIN = Models.Identifiers.GTIN(id)
            tradeitem.EPCISType = type

            // read the object
            ReadMasterDataObject(tradeitem, xTradeitem)
            doc.MasterData.add(tradeitem)
        }


        private fun ReadLocation(doc: EPCISBaseDocument, xLocation: XElement, type: String) {
            // read the GLN from the id
            val id = xLocation.attribute("id")?.value ?: ""
            val t = Setup.MasterDataTypes[type]
            val location = if (Activator.CreateInstance(t) is Location) {
                throw Exception("Failed to create instance of Location from type $t")
            } else {
                loc.GLN = Models.Identifiers.GLN(id)
                loc.EPCISType = type

                // read the object
                ReadMasterDataObject(loc, xLocation)
                doc.MasterData.add(loc)
            }
        }


        private fun ReadTradingParty(doc: EPCISBaseDocument, xTradingParty: XElement, type: String) {
            // read the PGLN from the id
            val id = xTradingParty.attribute("id")?.value ?: ""
            val tp = TradingParty()
            tp.PGLN = Models.Identifiers.PGLN(id)
            tp.EPCISType = type

            // read the object
            ReadMasterDataObject(tp, xTradingParty)
            doc.MasterData.add(tp)
        }


        private fun ReadUnknown(doc: EPCISBaseDocument, xVocabElement: XElement, type: String) {
            // read the PGLN from the id
            val id = xVocabElement.attribute("id")?.value ?: ""
            val ele = VocabularyElement()
            ele.ID = id
            ele.EPCISType = type

            // read the object
            ReadMasterDataObject(ele, xVocabElement)
            doc.MasterData.add(ele)
        }



        private fun ReadMasterDataObject(md: IVocabularyElement, xMasterData: XElement, readKDEs: Boolean = true) {
            val mappedProperties = OTMappingTypeInformation.GetMasterDataXmlTypeInfo(md.javaClass)

            // work on expanded objects...
            // these are objects on the vocab element represented by one or more attributes in the EPCIS master data
            val ignoreAttributes = mutableListOf<String>()
            for (property in mappedProperties.Properties.filter { it.Name == "" }) {
                val subMappedProperties = OTMappingTypeInformation.GetMasterDataXmlTypeInfo(property.Property.type)
                var setAttribute = false
                val subObject = property.Property.type.createInstance()
                if (subObject != null) {
                    for (xeAtt in xMasterData.elements("attribute")) {
                        val id = xeAtt.attribute("id")?.value ?: ""
                        val propMapping = subMappedProperties[id]
                        if (propMapping != null) {
                            if (!TrySetValueType(xeAtt.value, propMapping.property, subObject)) {
                                val value = ReadKDEObject(xeAtt, propMapping.property.type)
                                propMapping.property.setValue(subObject, value)
                            }
                            setAttribute = true
                            ignoreAttributes.add(id)
                        }
                    }
                    if (setAttribute) {
                        property.Property.setValue(md, subObject)
                    }
                }
            }

            // go through each standard attribute...
            for (xeAtt in xMasterData.elements("attribute")) {
                val id = xeAtt.attribute("id")?.value ?: ""

                if (ignoreAttributes.contains(id)) {
                    continue
                }

                val propMapping = mappedProperties[id]
                if (propMapping != null) {
                    if (!TrySetValueType(xeAtt.value, propMapping.property, md)) {
                        val value = ReadKDEObject(xeAtt, propMapping.property.type)
                        propMapping.property.setValue(md, value)
                    }
                } else if (readKDEs) {
                    if (xeAtt.hasElements) {
                        // serialize into object kde...
                        val kdeObject = MasterDataKDEObject("", id)
                        kdeObject.SetFromEPCISXml(xeAtt)
                        md.KDEs.add(kdeObject)
                    } else {
                        // serialize into string kde
                        val kdeString = MasterDataKDEString("", id)
                        kdeString.SetFromEPCISXml(xeAtt)
                        md.KDEs.add(kdeString)
                    }
                }
            }
        }


        private fun ReadKDEObject(xeAtt: XElement, t: Type): Any {
            val value = t.createInstance() ?: throw Exception("Failed to create instance of ${t.fullName}")

            if (value is MutableList<*>) {
                val list = value as MutableList<Any>
                for (xchild in xeAtt.elements()) {
                    val child = ReadKDEObject(xchild, t.genericTypeArguments[0])
                    list.add(child)
                }
            } else {
                // go through each property...
                for (p in t.declaredProperties) {
                    val xmlAtt = p.getCustomAttribute<OpenTraceabilityAttribute>()
                    if (xmlAtt != null) {
                        val x = xeAtt.element(xmlAtt.name)
                        if (x != null) {
                            val objAtt = p.getCustomAttribute<OpenTraceabilityObjectAttribute>()
                            if (objAtt != null) {
                                val o = ReadKDEObject(x, p.type)
                            } else if (!TrySetValueType(x.value, p, value)) {
                                throw Exception("Failed to set value type while reading KDE object. property = ${p.name}, type = ${t.fullName}, xml = ${x.toString()}")
                            }
                        }
                    }
                }
            }

            return value
        }


        private fun TrySetValueType(`val`: String, p: PropertyInfo, o: Any): Boolean {
            when {
                p.propertyType == String::class.java -> {
                    p.setValue(o, `val`)
                    return true
                }
                p.propertyType == MutableList::class.java && p.genericType == List::class.java -> {
                    val cur = p.getValue(o) as MutableList<String>?
                    if (cur == null) {
                        val newList = mutableListOf<String>()
                        newList.add(`val`)
                        p.setValue(o, newList)
                    } else {
                        cur.add(`val`)
                    }
                    return true
                }
                p.propertyType == Boolean::class.java || p.propertyType == java.lang.Boolean::class.java -> {
                    val v = java.lang.Boolean.parseBoolean(`val`)
                    p.setValue(o, v)
                    return true
                }
                p.propertyType == Double::class.java || p.propertyType == java.lang.Double::class.java -> {
                    val v = java.lang.Double.parseDouble(`val`)
                    p.setValue(o, v)
                    return true
                }
                p.propertyType == URI::class.java -> {
                    val v = URI.create(`val`)
                    p.setValue(o, v)
                    return true
                }
                p.propertyType == MutableList::class.java && p.genericType == List::class.java -> {
                    val l = mutableListOf<LanguageString>()
                    l.add(LanguageString("en-US", `val`))
                    p.setValue(o, l)
                    return true
                }
                p.propertyType == Country::class.java -> {
                    val v = Countries.Parse(`val`)
                    p.setValue(o, v)
                    return true
                }
                p.propertyType == PGLN::class.java -> {
                    val v = PGLN(`val`)
                    p.setValue(o, v)
                    return true
                }
                else -> return false
            }
        }

    }
}
